### ISSUE TITLE
Cross-references and numbering

### ISSUE THREAD

**Issue by jgm** [jgm/djot#30]
LaTeX has a flexible system for creating numbering counters, labels, and cross-references. This can be used with headings, tables, figures, equations, even list items. This is a must for serious academic writing, but it's not easy to see how to create a system that is sufficiently flexible but still natural for plain text writing and easy to use.

**Comment by jgm** [comment]
Pandoc's example lists offer some of the functionality of these, but only some. It would be better to have a more general solution, rather than implementing example lists here.

**Comment by uvtc** [comment]
Can you please give an example of what you mean by this? I'm familiar with putting an `#that-id` in a doc, and then linking to it from elsewhere with a link pointing to `#that-id`...

Also, where does "numbering" factor in here?

**Comment by jgm** [comment]
In LaTeX you can have
```
\section{My section heading}\label{mysec}
```
and then at some other point in the document,
```
See Section \ref{mysec}.
```
The section will be automatically numbered and the reference will resolve automatically.  You can do the same kind of thing for other automatically numbered things, like equations, figures, and tables.

**Comment by jgm** [comment]
pandoc-crossref provides these capabilities as an add-on to pandoc.

**Comment by uvtc** [comment]
Thanks, John. I found the link to [pandoc-crossref](https://github.com/lierdakil/pandoc-crossref).

Although I don't understand your "mysec" example (since you're explicitly naming the reference "mysec" and there doesn't appear to be any numbering or auto-numbering involved), I can see how, if you have a bunch of display-math examples, they would/should be auto-numbered... though I'm not sure how I'd know what number they would be (or what their `#id`'s would be to link to them).

**Comment by stagnation** [comment]
Nice, I use this all the time in my own notes (mostly in rst but not sphinx nor pandoc) and would often like to write documents with links to sections in the same document, and sometimes with references to sections in other documents. There one would need some kind of name spaces I think, sphinx "extends" rst to add that, but with a global namespace, so documents add anchor labels with unique names for sections that are linked to.

**Comment by gfarrell** [comment]
I just stumbled across djot as I've been implementing my own sortof-markdownish-thing for writing essays to publish online, for which I've been trying to implement three things missing in markdown which I miss from LaTeX:

* footnotes (I see djot has these already)
* citations (there's some discussion in #32, but I don't think this was ever implemented?)
* references with automated numbering (for figures, tables, sections, etc.)

So, since this issue is about references, @jgm are you intending to work this into djot or is there more discussion to be had?

I came up with the same syntax as you for footnotes `^[...]` (but with some differences which are off-topic here), for citations I like the ideas in #32, but I treated citations as an extension of the concept of a reference, for which I used `@{type:label}` syntax.

Reference anchors, in my design, were part of the blocks, so:

```
\`\`\`{label}haskell
-- some code
\`\`\`

![a nice sunset][sunset-img]{sunset}
```

And I limited the types of reference to sections, citations, figures, tables, and (code-)listings, which I know is restrictive but it prevented me from having to use a convention that I used in my LaTeX docs of `\ref{fig:cool-graph}` and `\ref{table:amazing-results}` which was easy to get wrong.

Anyway, the above is just _stuff I tried_ but I'd be interested in trying to get citations and references into djot (and djoths) because I think I'd rather use your implementation for my website than have to solve all the problems you've already solved building my own markup language. Is there something I can do?

**Comment by jgm** [comment]
At this point we're still at the design phase, figuring out what syntax for citations and references should be. Implementing the decisions will be relatively straightforward.

**Comment by Omikhleia** [comment]
FWIW, regarding cross-references, in my workflows with either Markdown or Djot inputs, I went for "filtering" empty links (i.e. without link text)[^2]:
- `[](#xxx)` refers to whatever identifier `xxx` points at, by "closest" numbering -- so that could be a figure, table, section or any other numbering scheme there is at that point (equation number, verse number in poetry, etc.)
- (Despite using English keywords not being perfect...) I use (pseudo-)classes when I want :
  - The page reference `[](#xxx){.page}`
  - The upper section title `[]{#xxx){.title}`
  - Or the upper sectioning number `[](#xxx){.section}` (in the broad sense, i.e. the closest section, chapter, etc. in which the item appears)

The syntax is not that friendly, but has the advantage of remaining somehow readable (and avoiding low-level syntax extensions, being entirely done via a filter and/or by the renderer itself, which is responsible anyway for tracking the underlying counters[^1]).


[^2]: Assuming these normally don't occur naturally in input files since they not produce anything very visible ^^

[^1]: I am using SILE, not LaTeX, but at the core this boils down to the same kind of `\label` and `\ref`/`\pageref` logic.

**Comment by jgm** [comment]
LaTeX allows quite a bit of flexibility in its counters. For example, you might want the figures to be numbered after the section or chapter, e.g. Figure 3.2.  I think this kind of configurability would be necessary for a really workable solution.

**Comment by Omikhleia** [comment]
> n implementation, both provide reasonable defaults for label-prefix (eg, fig:) and output prefix (eg, Figure). I think these are good decisions

That a renderer styling issue no? (For different current document languages and usages, independent from the Djot syntax)

**Comment by olynch** [comment]
Here's a possible design.

1. As in pandoc-citeproc and quarto, re-use the existing id system for references. So `\label{fig:pig}` would just be

```
{#fig:pig}
![](pig.png)
```

2. Separate counters from references. Counters are accessed via some syntax like `!thm`.

3. Allow a block to set its own "cleverref" name with something like `[@ Theorem !thm]` (this syntax I don't love, but something like that).

So a figure would look something like

```
{#fig:pig}
{.figure}
:::
![](pig.png)

*[@ Figure !fig].* A pig.
:::

In @fig:pig, we see that...
```

and `@fig:pig` would resolve to `[Figure 1](#fig:pig)`.

4. Allow "custom resolvers" for `@` references, so that users can plug in, say, something that reads a `.bib` file and resolves `@maclane-1971-categories` to `[Mac Lane [1]](#maclane-1971-categories)`.  Note that this could equally well be used for resolving `@jgm` to `[John Macfarlane](https://github.com/jgm)` in a github comment, if github used djot...

**Comment by jgm** [comment]
Those are reasonable suggestions.  Another issue is how one would control numbering styles.  One might want figures numbered 1.2, 1.3, etc.; one might want different numbering schemes for sections or equations; etc.  In LaTeX you can control all of this.

**Comment by olynch** [comment]
It seems like it would be against the philosophy of djot to have some kind of domain-specific language for setting up counters. Perhaps, just like with resolvers for `@`, we could also have resolvers for `!`, so that deciding how the `!fig` interacts with sections/subsections would be done via passing a function into the config for the renderer. And the default resolver would just use a "flat" counter.

Of course, one way of configuring these resolvers could be via the frontmatter metadata in a static site generator, but this would just be one way; this doesn't have to be in the djot spec.

**Comment by olynch** [comment]
Thinking about this a bit more.

I think that a good name for the `!fig` syntax would be "stamp" as in

<img width="300" height="300" alt="automatic numbering stamp" src="https://github.com/user-attachments/assets/6180a1f4-6477-4584-bd17-df197ef7c23e" />

The mnemonic would be that it "stamps" the current block, both visually (in fact, one could configure `!fig` to expand out to `Figure 2.`) and logically (setting what the name used for `@`-references would be). I think the `[@ Theorem !thm]` syntax from before is unnecessary; this should just be part of the configuration for stamps.

The configuration for a particular stamp might say that applying the stamp reads/writes from one or more counters (which would allow Figure 1.2, etc.), but there doesn't need to be a syntax in djot for "raw counters".

And then there could very well be a setting called "auto-stamp section headings" which would treat

```
## Pigs of paradise
```

as

```
## !h2 Pigs of paradise
```

though (importantly) it would keep the automatically generated section id as `Pigs-of-paradise`, not `1-Pigs-of-paradise`

So to summarize, the only changes to the grammar would be adding inline nodes for `@fig:pig` and `!fig`, and then implementations could add "hooks" for interpreting those.

**Comment by jgm** [comment]
So, to make sure I understand:

- `@fig:pig` resolves to the number of the associated figure labeled `#fig:pig`
- `!fig:pig` resolves to this number, together with the label `Figure` (configurable), and produces a link

Note that @fig:pig overlaps with proposed citation syntax.  That could be a problem, or it could be okay, perhaps.

**Comment by olynch** [comment]
Ah, not quite.

`!fig` *stamps* the current block using the `fig` stamp, which is configured to output `Figure 1.` and associate the id of the current block (`fig:pig`) with the name `Figure 1.`. The fact that the id is `fig:pig` has no relationship to the `fig` stamp, it's just following the naming convention I often use for labels in latex. Also note that one might wish to output something different than the name you associate the block with, as in the case of section headers where you output `1.` but then associate the id with `Section 1.`.

Then, later on, we can use `@fig:pig` to retrieve that name (and link it to the id).

The `@` syntax would be more general than just refs or citations; a programmatic user of djot could attach a custom function to a hook which would allow resolving an `@` reference with whatever custom logic they want. So for instance, instead of using a `.bib` file the renderer could talk directly to Zotero, or to a database if djot were used in a wiki, etc.

One advantage of this system is that we don't need a concept of "environment" like in latex; `!fig` can stamp anything, it is just a good idea to use it to stamp a div with the HTML class of `figure`.

**Comment by olynch** [comment]
Similarly, if multiple djot files were used for the chapters of a book, stamps could be persistent across chapters, and in theory this would allow rebuilding a single chapter without rebuilding the whole book. And "non-displayed" stamps could be used to tag things by subject to build indices.

But again, this wouldn't be part of the djot spec, these are just different ways that programs could hook into the djot syntax.

**Comment by olynch** [comment]
Any thoughts on this @jgm? I'd be happy to try and implement a prototype in djoths.

**Comment by jgm** [comment]
```
*[@ Figure !fig].* A pig.
```
I didn't understand whether this was supposed to be optional.  Arguably, generation of the label "Figure 1." should be part of a rendering style, and not something you should need to specify.

If we don't want to specify it manually, then I'm not sure why we'd want !fig as an inline element. One could instead use a block attribute, e.g. `series=figure` or perhaps `!fig` could be shorthand for that in attributes....

**Comment by olynch** [comment]
That's a good point! I was envisioning it inline to make it more minimal, but perhaps you are right that it works better as part of the rendering style. Another option is that the label "Figure 1." could be what is produced by `!fig`, so it would be

!fig. A pig

This would allow for customizing where in the block the label shows up. I could imagine either block attribute or inline both being reasonable approaches though.

Maybe I can try implementing the "stamp database" in djot-hs, and then we could play around with different syntax that hooked into it.

**Comment by jgm** [comment]
Doesn't hurt to experiment.
